File: TheBossInJS/City.js
var ACTION_TYPE = require("./action_type");
var CITY_ID = require("./city_id");
var ActionFactory = require("./action");

function City(id, createActions) {

    this.id = id;
    this.gangsterMap = {}; // map from Boss to gangster list[]
    this.unShowenActions = [];

    // init
    createActions.call(this);
    this.nextHand();
}

City.prototype.toString = function() {
    var str = "/" + CITY_ID.toString(this.id) + "/\n";
    for (var i in this.actions) {
        str += ("\t" + this.actions[i] + "\n");
    }
    return str;
}

City.prototype.nextHand = function() {

    this.unShowenActions.length = 0; // empty array
    this.unShowenActions = this.actions.slice();
    //console.log("next hand:, unShowenActions = " + this.unShowenActions);

    // empty occupied gangsters;
    for (var i in this.gangsterMap)
        this.gangsterMap[i].length = 0;
}

City.prototype.removeProtector = function() {

    for (var i = 0; i< this.actions.length; i++) {
        var a = this.actions[i];
        if (a.type == ACTION_TYPE.REWARD)
            a.removeRewardListener();
    }
}

City.prototype.addProtector = function(protector) {

    for (var i = 0; i< this.actions.length; i++) {
        var a = this.actions[i];
        if (a.type == ACTION_TYPE.REWARD)
            a.addRewardListener(protector);
    }
}

City.prototype.occupy = function(boss, gangster) {

    if (!this.gangsterMap[boss])
        this.gangsterMap[boss] = [];
    this.gangsterMap[boss].push(gangster);
}

City.prototype.showAction = function(action) {

    console.log("Show Action: " + action.toString());
    this.unShowenActions.splice(this.unShowenActions.indexOf(action), 1);
}

module.exports.PROTECTOR = new City(
        CITY_ID.CHICAGO, 
        function() { this.actions = [];}
);

module.exports.PROTECTOR.reward = function(money) {

    var action = ActionFactory.createAction(
                    ACTION_TYPE.REWARD,
                    this.id,
                    Math.floor(money/2));
    action.execFn(this.gangsterMap);
}

module.exports.CITIES = [
    new City(CITY_ID.NEW_YORK, function() {
        this.actions = [];
        for (var i = 0; i <= 3; i++) {
            this.actions.push(
                ActionFactory.createAction(
                    ACTION_TYPE.REWARD, this.id, i));
        }
    }),

    new City(CITY_ID.BOSTON, function() {
        this.actions = [];
        for (var i = 0; i < 2; i ++) {
            this.actions.push(
                ActionFactory.createAction(
                    ACTION_TYPE.REWARD, this.id, 3));
        }
        this.actions.push(
            ActionFactory.createAction(
                ACTION_TYPE.IMPRISON, this.id));
        this.actions.push(
            ActionFactory.createAction(
                ACTION_TYPE.HOSPITALIZE, this.id));
    }),

    new City(CITY_ID.DETROIT, function() {
        this.actions = [];
        for (var i = 1; i <= 4; i++) {
            this.actions.push(
                ActionFactory.createAction(
                    ACTION_TYPE.REWARD, this.id, i));
        }
        this.actions.push(
            ActionFactory.createAction(
                ACTION_TYPE.SHOOT, this.id));
    }),

    new City(CITY_ID.KANSAS_CITY, function() {
        this.actions = [];
        for (var i = 2; i <= 3; i++) {
            this.actions.push(
                ActionFactory.createAction(
                    ACTION_TYPE.REWARD, this.id, i));
        }
        this.actions.push(
            ActionFactory.createAction(
                ACTION_TYPE.SHOOT, this.id));
    }),

    new City(CITY_ID.CINCINNATI, function() {
        this.actions = [];
        for (var i = 1; i <= 3; i ++) {
            this.actions.push(
                ActionFactory.createAction(
                    ACTION_TYPE.REWARD, this.id, i));
        }
        this.actions.push(
            ActionFactory.createAction(
                ACTION_TYPE.BANISH, this.id));
    }),

    new City(CITY_ID.MEMPHIS, function() {
        this.actions = [];
        for (var i = 1; i <= 3; i ++) {
            this.actions.push(
                ActionFactory.createAction(
                    ACTION_TYPE.REWARD, this.id, i));
        }
        this.actions.push(
            ActionFactory.createAction(
                ACTION_TYPE.HOSPITALIZE, this.id));
    }),

    new City(CITY_ID.PHILADELPHIA, function() {
        this.actions = [];
        this.actions.push(
            ActionFactory.createAction(
                ACTION_TYPE.REWARD, this.id, 3));
        this.actions.push(
            ActionFactory.createAction(
                ACTION_TYPE.SHOOT, this.id));
        this.actions.push(
            ActionFactory.createAction(
                ACTION_TYPE.IMPRISON, this.id));
    })
];

File: TheBoss/src/model/City.java
package model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import model.constant.ACTION_TYPE;
import model.constant.CITY_ID;
import model.impl.Boss;
import model.impl.Gangster;
import model.impl.action.Reward;


public abstract class City implements Serializable {

	private static final long serialVersionUID = 1L;
	
	protected CITY_ID id;
	protected Map<Boss, List<Gangster>> gangsterMap;
	protected List<CardAction> actions, unShowenActions;
	
	public City(CITY_ID id) {
		
		this.id = id;
		gangsterMap = new HashMap<Boss, List<Gangster>>(
							TheBossBoardGame.MAX_NUM_OF_BOSSES);
		actions = new ArrayList<CardAction>();
		unShowenActions = new ArrayList<CardAction>();
		
		createActions();
		nextHand();
	}
	
	protected abstract void createActions();
	
	public CardAction[] getActions(){
		
		return actions.toArray(new CardAction[actions.size()]);
	}
	
	public void nextHand() {
		
		unShowenActions.clear();
		unShowenActions.addAll(actions);
		
		for(List<Gangster> g: gangsterMap.values())
			g.clear();
		gangsterMap.clear();
	}
	
	public void removeProtector() {
	
        for(CardAction a: actions){
            if(a.getType() == ACTION_TYPE.REWARD)
                ((Reward) a).removeRewardListener();
        }
	}
	
	public void addProtector(RewardListener protector) {
	
		// need to give some fee to the protector...
		for(CardAction a: actions){
			if(a.getType() == ACTION_TYPE.REWARD)
				((Reward) a).addRewardListener(protector);
		}
	}
	
	public void occupy(Boss boss, Gangster gangster) {
		
		if(!gangsterMap.containsKey(boss)){
			gangsterMap.put(boss, new ArrayList<Gangster>());
		}
		gangsterMap.get(boss).add(gangster);
	}
	
	public void showAction(CardAction actionShown) {
		
		unShowenActions.remove(actionShown);
	}
	
	public CardAction winningOrSanction() {
		
		if(unShowenActions.size() != 1)
			throw new RuntimeException("unShwonActions.size shall be 1");
		
		CardAction lastAction = unShowenActions.get(0);
		lastAction.execute(gangsterMap);
		
		return lastAction;
	}
	
	@Override
	public String toString() {
		
		return id.toString();
	}
	
	// for debug
	public void printGangsterMap() {

        System.out.println("====" + this + "====");
	    for (Boss b: gangsterMap.keySet()) {
	        List<Gangster> g = gangsterMap.get(b);
	        System.out.println(b.getId() + "(" + b.getName() + ") occupies " + g.size() + " gangsters.");
	    }
	}
}
